internal/worker/task_handler.go

```diff
@@ -152,6 +154,51 @@ func (t *TaskHandler) setNextBootDevice(ctx context.Context, bmc device.Queryor)
 	return t.successful(ctx, "next boot device set successfully: "+t.task.Parameters.ActionParameter)
 }
 
+func (t *TaskHandler) validateFirmware(ctx context.Context, bmc device.Queryor) error {
+	t.loggerEntry.Info("starting firmware validation")
+
+	deadline := time.Now().Add(time.Duration(t.task.Parameters.ValidateFirmwareTimeout * uint(time.Second)))
+
+	// First reboot the BMC to ensure it's running the desired firmware
+	if err := bmc.PowerCycleBMC(ctx); err != nil {
+		return t.failedWithError(ctx, "failed to power cycle BMC", err)
+	}
+
+	var err error
+
+	// Next we want to cycle the host, but the BMC will take some
+	// time to reboot, so retry once every 30 seconds up to our
+	// timeout deadline (ideally we'd have a way to distinguish
+	// failures that are due to the BMC not being back online yet
+	// from ones that aren't going to be resolved by waiting and
+	// retrying...)
+	for time.Now().Before(deadline) {
+		time.Sleep(30 * time.Second)
```

Comment:
if you move this down a few lines, you will never sleep in the happy path, which I think is desired?

Reply:
This gets into the raciness alluded to in the PR description...if we check immediately I suspect we'd be pretty likely to see a "booted successfully" POST code left over from the current/previous boot cycle; ordering it this way puts a delay in between that would hopefully give it time to update its POST state after the reset.  (Not the most bulletproof technique in the world, but without a totally different approach I'm not sure at the moment how we can do substantially better.)

Reply:
This is in its own thread, cf. [here](https://github.com/metal-toolbox/ctrl/blob/8d903a237aa6789b873d55de735434818237e53e/controller_nats.go#L288).